doc/user-manual/language/pragmas.lagda.rst

```diff
@@ -128,6 +130,37 @@ Example::
 Aside from datatypes, this pragma can also be used to mark other
 definitions as being injective (for example postulates).
 
+.. _injective-for-inference-pragma:
+
+The ``INJECTIVE_FOR_INFERENCE`` pragma
+______________________________________
+
+Treats functions as injective for type inference. This behaves like a
+local version of :option:`--lossy-unification` and has the same
+potential issues. Since Agda can not always infer whether a function
+is injective it can be used to get stronger unification for those
+functions.
+
+Example::
+
+  open import Agda.Builtin.Equality
+  open import Agda.Builtin.Nat
+  open import Agda.Builtin.Sigma
+
+  _×_ : Set → Set → Set
+  A × B = Σ A (λ _ → B)
+
+  f : {A : Set} → Nat × A → Nat × A
+  f (n , a) = (0 , a)
```

Comment:
Is the intention to mainly use this pragma with injective functions? I'm still waiting for some motivation for this new feature. If this was discussed during the recent Agda meeting, please include a summary here.

Reply:
My main use ATM is to use it with functions that are injective in a setoid. It's a bit hard to find good self-contained examples for injective functions, since postulates/module parameters are already treated as being injective. Here's one example I could find - imagine the double reverse being something more interesting.
``` agda
module _ {A : Set} where
  _++_ : List A → List A → List A
  []       ++ ys = ys
  (x ∷ xs) ++ ys = x ∷ (xs ++ ys)

  reverse : List A → List A
  reverse []      = []
  reverse (x ∷ l) = reverse l ++ (x ∷ [])

  {-# INJECTIVE_FOR_INFERENCE reverse #-}

  reverse-cong : {l l' : List A} → reverse l ≡ reverse l' → reverse (reverse l) ≡ reverse (reverse l')
  reverse-cong h rewrite h = refl

  reverse-cong' : {l l' : List A} → reverse l ≡ reverse l' → reverse (reverse l) ≡ reverse (reverse l')
  reverse-cong' h = reverse-cong h
```